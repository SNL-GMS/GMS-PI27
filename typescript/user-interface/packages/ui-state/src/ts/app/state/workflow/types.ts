import type { CommonTypes, WorkflowTypes } from '@gms/common-model';
import type { Nullable } from '@gms/common-model/lib/type-util/type-util';

export interface WorkflowState {
  timeRange: Nullable<CommonTypes.TimeRange>;
  stationGroup: WorkflowTypes.StationGroup;
  openIntervalName: string; // e.x AL1
  openActivityNames: string[]; // e.x Event Review
  analysisMode: WorkflowTypes.AnalysisMode | null;
}
