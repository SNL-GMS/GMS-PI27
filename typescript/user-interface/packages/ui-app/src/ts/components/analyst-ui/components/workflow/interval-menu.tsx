import { Menu, MenuItem } from '@blueprintjs/core';
import type { WorkflowTypes } from '@gms/common-model';
import {
  isActivityInterval,
  isInteractiveAnalysisStageInterval,
  isStageInterval
} from '@gms/common-model/lib/workflow/types';
import type { ImperativeContextMenuProps } from '@gms/ui-core-components';
import { showImperativeReduxContextMenu, useAppSelector } from '@gms/ui-state';
import flatMap from 'lodash/flatMap';
import React from 'react';

import { getWorkflowPanelElement } from './workflow-util';

export interface IntervalMenuProps {
  readonly interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval;
  readonly isSelectedInterval: boolean;
  readonly allActivitiesOpenForSelectedInterval: boolean;

  readonly openCallback: (
    interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval
  ) => void;
  readonly closeCallback: (
    interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval
  ) => void;
}

/**
 * Component that renders the interval context menu.
 */
export function IntervalMenu(props: IntervalMenuProps) {
  const openIntervalName = useAppSelector(state => state.app.workflow.openIntervalName);
  const {
    interval,
    isSelectedInterval,
    allActivitiesOpenForSelectedInterval,
    openCallback,
    closeCallback
  } = props;

  const userName = useAppSelector(state => state.app.userSession.authenticationStatus.userName);

  const isDisabled =
    (allActivitiesOpenForSelectedInterval &&
      isSelectedInterval &&
      openIntervalName === interval.intervalId.definitionId.name) ||
    (isSelectedInterval && interval.intervalId.definitionId.name !== openIntervalName);

  // determine if the logged in user is an active analyst; if so allow them to close that interval
  let isActiveAnalyst = false;
  if (isStageInterval(interval)) {
    if (isInteractiveAnalysisStageInterval(interval)) {
      isActiveAnalyst = flatMap(interval.activityIntervals.map(a => a.activeAnalysts)).includes(
        userName
      );
    }
  } else if (isActivityInterval(interval)) {
    isActiveAnalyst = interval.activeAnalysts.includes(userName);
  }

  return (
    <Menu>
      <MenuItem
        className="menu-item-open-interval"
        data-cy="open-interval-btn"
        text={isActiveAnalyst && !isDisabled ? 'Reopen interval' : 'Open interval'}
        disabled={isDisabled}
        onClick={() => openCallback(interval)}
      />
      <MenuItem
        className="menu-item-close-interval"
        data-cy="close-interval-btn"
        text="Close interval"
        disabled={!isSelectedInterval && !isActiveAnalyst}
        onClick={() => closeCallback(interval)}
      />
    </Menu>
  );
}

/**
 * Shows the {@link IntervalMenu} context Menu.
 *
 * @param event the event that invoked the action
 * @param props the {@link IntervalMenuProps} props
 * @param options (optional) imperative context menu options
 */
export const showIntervalMenu = (
  event: React.MouseEvent | MouseEvent,
  props: IntervalMenuProps,
  options: Pick<ImperativeContextMenuProps, 'activeElementOnClose' | 'onClose'> = {
    activeElementOnClose: getWorkflowPanelElement(),
    onClose: undefined
  }
) => {
  const {
    allActivitiesOpenForSelectedInterval,
    closeCallback,
    interval,
    isSelectedInterval,
    openCallback
  } = props;
  event.preventDefault();
  showImperativeReduxContextMenu({
    content: (
      <IntervalMenu
        allActivitiesOpenForSelectedInterval={allActivitiesOpenForSelectedInterval}
        closeCallback={closeCallback}
        interval={interval}
        isSelectedInterval={isSelectedInterval}
        openCallback={openCallback}
      />
    ),
    ...options,
    targetOffset: {
      left: event.clientX,
      top: event.clientY
    }
  });
};
