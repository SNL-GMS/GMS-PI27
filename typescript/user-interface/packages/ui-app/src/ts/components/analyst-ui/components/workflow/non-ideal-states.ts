import { Endpoints } from '@gms/common-model';
import { capitalizeFirstLetters } from '@gms/common-util';
import type { NonIdealStateDefinition } from '@gms/ui-core-components';
import {
  nonIdealStateWithError,
  nonIdealStateWithNoSpinner,
  nonIdealStateWithSpinner
} from '@gms/ui-core-components';
import type { StageIntervalsByIdAndTimeQuery, WorkflowQuery } from '@gms/ui-state';

export const workflowQueryNonIdealStates: NonIdealStateDefinition<unknown>[] = [
  {
    condition: (props: { workflowQuery: WorkflowQuery }): boolean => {
      return props.workflowQuery?.isLoading;
    },
    element: nonIdealStateWithSpinner('Loading', 'Workflow Data')
  },
  {
    condition: (props: { workflowQuery: WorkflowQuery }): boolean => {
      return (
        props.workflowQuery?.isError &&
        (props.workflowQuery?.error?.message?.includes('404') ||
          props.workflowQuery?.error?.message?.includes('503'))
      );
    },
    element: nonIdealStateWithError(
      'Error',
      `Unable to communicate with ${capitalizeFirstLetters(
        Endpoints.WorkflowManagerServiceUrls.workflow.friendlyName
      )} Service`
    )
  },
  {
    condition: (props: { workflowQuery: WorkflowQuery }): boolean => {
      return props.workflowQuery?.isError;
    },
    element: nonIdealStateWithError('Error', 'Problem loading workflow data')
  },
  {
    condition: (props: { workflowQuery: WorkflowQuery }): boolean => {
      return (
        !props.workflowQuery?.isLoading &&
        !props.workflowQuery?.isError &&
        !props.workflowQuery?.data &&
        props.workflowQuery?.data.stages.length === 0
      );
    },
    element: nonIdealStateWithNoSpinner('No Data', 'Workflow stages returned empty', 'exclude-row')
  }
];

export const workflowIntervalQueryNonIdealStates: NonIdealStateDefinition<unknown>[] = [
  {
    condition: (props: {
      readonly hasFetchedInitialIntervals: boolean;
      workflowIntervalQuery: StageIntervalsByIdAndTimeQuery;
    }): boolean => {
      // only show the loading state if the on the initial fetch of the interval data
      return !props.hasFetchedInitialIntervals && props.workflowIntervalQuery?.isLoading;
    },
    element: nonIdealStateWithSpinner('Loading', 'Workflow Interval Data')
  },
  {
    condition: (props: { workflowIntervalQuery: StageIntervalsByIdAndTimeQuery }): boolean => {
      return (
        !props.workflowIntervalQuery?.isLoading &&
        !props.workflowIntervalQuery?.isError &&
        (!props.workflowIntervalQuery.data || props.workflowIntervalQuery.data.length === 0)
      );
    },
    element: nonIdealStateWithNoSpinner('No Data', 'Stage intervals returned empty', 'exclude-row')
  },
  {
    condition: (props: { workflowIntervalQuery: StageIntervalsByIdAndTimeQuery }): boolean => {
      return props.workflowIntervalQuery?.isError;
    },
    element: nonIdealStateWithError('Error', 'Problem loading workflow interval data')
  }
];
