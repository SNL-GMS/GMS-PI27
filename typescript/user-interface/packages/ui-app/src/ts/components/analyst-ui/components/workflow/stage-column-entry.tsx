import type { WorkflowTypes } from '@gms/common-model';
import React from 'react';

import { SequenceIntervalCell } from './sequence-interval-cell';

export interface StageColumnEntryProps {
  readonly stageInterval: WorkflowTypes.StageInterval;
  readonly workflow: WorkflowTypes.Workflow | undefined;
}

function StageColumnEntryComponent(props: React.PropsWithChildren<StageColumnEntryProps>) {
  const { stageInterval, workflow } = props;
  return (
    <SequenceIntervalCell
      stageInterval={stageInterval}
      key={stageInterval.intervalId.definitionId.name}
      workflow={workflow}
    />
  );
}

/**
 * Row entry wrapper for Sequence cells
 */
export const StageColumnEntry: React.FunctionComponent<StageColumnEntryProps> =
  React.memo(StageColumnEntryComponent);
