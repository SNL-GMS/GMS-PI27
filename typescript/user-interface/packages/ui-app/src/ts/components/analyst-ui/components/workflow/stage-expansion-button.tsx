import { Button } from '@blueprintjs/core';
import React from 'react';

export interface StageExpansionButtonProps {
  readonly isExpanded: boolean;
  readonly disabled: boolean;
  readonly stageName: string;
  readonly toggle: () => void;
}

function StageExpansionButtonComponent(props: StageExpansionButtonProps) {
  const { disabled, stageName, isExpanded, toggle } = props;
  const smallSign = isExpanded ? 'small-minus' : 'small-plus';
  return !disabled ? (
    <Button
      key={stageName}
      className="stage-row__expand-button"
      icon={smallSign}
      onClick={toggle}
      disabled={disabled}
    />
  ) : null;
}

export const StageExpansionButton: React.FunctionComponent<StageExpansionButtonProps> = React.memo(
  StageExpansionButtonComponent
);
