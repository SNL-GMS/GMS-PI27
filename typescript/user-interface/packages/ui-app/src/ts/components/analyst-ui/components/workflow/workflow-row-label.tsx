import classNames from 'classnames';
import React from 'react';

export interface WorkflowRowLabelProps {
  readonly label: string;
  readonly isActivityRow: boolean;
}

export function WorkflowRowLabel(props: WorkflowRowLabelProps) {
  const { label, isActivityRow } = props;
  return (
    <div
      key={label}
      data-cy={`workflow-table-label-${label}`}
      className={classNames('workflow-table-label', {
        'workflow-table-label--activity': isActivityRow
      })}
    >
      <div>{label}</div>
    </div>
  );
}
