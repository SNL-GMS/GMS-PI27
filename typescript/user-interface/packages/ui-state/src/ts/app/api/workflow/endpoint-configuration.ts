import { Endpoints } from '@gms/common-model';
import { UI_URL } from '@gms/common-util';
import type { RequestConfig, ServiceDefinition } from '@gms/ui-workers';

import { createServiceDefinition } from '../create-service-definition';
import { prioritizeRequests } from '../request-priority';

/**
 * Workflow request config definition
 */
export interface WorkflowRequestConfig extends RequestConfig {
  readonly workflow: {
    readonly baseUrl: string;
    readonly services: {
      readonly workflow: ServiceDefinition;
      readonly stageIntervalsByIdAndTime: ServiceDefinition;
      readonly updateActivityIntervalStatus: ServiceDefinition;
      readonly updateStageIntervalStatus: ServiceDefinition;
    };
  };
}

const baseUrl = `${UI_URL}${Endpoints.WorkflowManagerServiceUrls.baseUrl}`;

/**
 * The workflow request config for all services.
 */
export const config: WorkflowRequestConfig = {
  workflow: {
    baseUrl,
    services: prioritizeRequests({
      workflow: createServiceDefinition({
        baseUrl,
        url: Endpoints.WorkflowManagerServiceUrls.workflow.url,
        friendlyName: Endpoints.WorkflowManagerServiceUrls.workflow.friendlyName,
        accept: 'application/json',
        contentType: 'text/plain',
        data: `"PlaceHolder"`
      }),
      stageIntervalsByIdAndTime: createServiceDefinition({
        baseUrl,
        url: Endpoints.WorkflowManagerServiceUrls.stageIntervalsByIdAndTime.url,
        friendlyName: Endpoints.WorkflowManagerServiceUrls.stageIntervalsByIdAndTime.friendlyName,
        accept: 'application/json'
      }),
      updateActivityIntervalStatus: createServiceDefinition({
        baseUrl,
        url: Endpoints.WorkflowManagerServiceUrls.updateActivityIntervalStatus.url,
        friendlyName:
          Endpoints.WorkflowManagerServiceUrls.updateActivityIntervalStatus.friendlyName,
        accept: 'application/json'
      }),
      updateStageIntervalStatus: createServiceDefinition({
        baseUrl,
        url: Endpoints.WorkflowManagerServiceUrls.updateStageIntervalStatus.url,
        friendlyName: Endpoints.WorkflowManagerServiceUrls.updateStageIntervalStatus.friendlyName
      })
    })
  }
};

export type WorkflowServices = keyof typeof config.workflow.services;
