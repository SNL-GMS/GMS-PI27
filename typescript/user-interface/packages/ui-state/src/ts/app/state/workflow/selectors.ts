import type { TimeRange } from '@gms/common-model/lib/common/types';
import type { Nullable } from '@gms/common-model/lib/type-util/type-util';
import type { AnalysisMode } from '@gms/common-model/lib/workflow/types';
import { createSelector } from '@reduxjs/toolkit';

import type { AppState } from '../../store';

/**
 * A redux selector for returning the open interval name.
 *
 * @example const name = useAppState(selectOpenIntervalName);
 *
 * @param state the redux app state
 * @returns the open interval name
 */
export const selectOpenIntervalName = (state: AppState): string =>
  state.app.workflow.openIntervalName;

/**
 * A redux selector for returning the open activity names.
 *
 * @example const names = useAppState(selectOpenActivityNames);
 *
 * @param state the redux app state
 * @returns the list of open activity names
 */
export const selectOpenActivityNames = (state: AppState): string[] =>
  state.app.workflow.openActivityNames;

/**
 * A redux selector for returning the open time range.
 *
 * @example const timerange = useAppState(selectWorkflowTimerange);
 *
 * @param state the redux app state
 * @returns the open time range
 */
export const selectWorkflowTimeRange = (state: AppState): Nullable<TimeRange> =>
  state.app.workflow.timeRange;

/**
 * A redux selector for returning the analysis mode
 *
 * @example const analysisMode = useAppState(selectWorkflowAnalysisMode);
 *
 * @param state the redux app state
 * @returns the analysis mode
 */
export const selectWorkflowAnalysisMode = (state: AppState): AnalysisMode | null =>
  state.app.workflow.analysisMode;

/**
 * Create a unique key based on the selected Workflow Interval
 *
 * @returns string to be used as a unique ID
 */
export const selectWorkflowIntervalUniqueId: (state: AppState) => string = createSelector(
  [selectWorkflowTimeRange, selectWorkflowAnalysisMode, selectOpenIntervalName],
  (currentInterval, analysisMode, openIntervalName) => {
    return `currentIntervalStartTime ${currentInterval.startTimeSecs} analysisMode ${analysisMode} openIntervalName ${openIntervalName}`;
  }
);

/**
 * A redux selector for returning the station group
 *
 * @example const stationGroup = useAppState(selectStationGroup);
 *
 * @param state the redux app state
 * @returns the station group
 */
export const selectStationGroup = (state: AppState) => state.app.workflow.stationGroup;

/**
 * A redux selector for returning the time range
 *
 * @example const timeRange = useAppState(timeRange);
 *
 * @param state the redux app state
 * @returns the time range
 */
export const selectTimeRange = (state: AppState) => state.app.workflow.timeRange;
