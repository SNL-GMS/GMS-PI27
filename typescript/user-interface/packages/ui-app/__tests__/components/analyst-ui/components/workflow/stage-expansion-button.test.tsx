/* eslint-disable react/jsx-props-no-spreading */
import { fireEvent, render, screen } from '@testing-library/react';
import * as React from 'react';

import type { StageExpansionButtonProps } from '../../../../../src/ts/components/analyst-ui/components/workflow/stage-expansion-button';
import { StageExpansionButton } from '../../../../../src/ts/components/analyst-ui/components/workflow/stage-expansion-button';

// set up window alert and open so we don't see errors
window.alert = jest.fn();
window.open = jest.fn();

window.ResizeObserver = jest.fn(() => {
  return { observe: jest.fn(), disconnect: jest.fn(), unobserve: jest.fn() };
});
const globalAny: any = global;
globalAny.ResizeObserver = window.ResizeObserver;
globalAny.DOMRect = jest.fn(() => ({}));

const MOCK_TIME = 1606818240000;
global.Date.now = jest.fn(() => MOCK_TIME);

describe('Stage Expansion Button', () => {
  it('is exported', () => {
    expect(StageExpansionButton).toBeDefined();
  });

  const stageExpansionButtonProps: StageExpansionButtonProps = {
    isExpanded: false,
    disabled: false,
    stageName: 'stage',
    toggle: jest.fn()
  };

  it('matches snapshot', () => {
    const component = render(<StageExpansionButton {...stageExpansionButtonProps} />);
    expect(component.baseElement).toMatchSnapshot();
  });

  it('Stage Expansion Button functions and clicks', () => {
    render(<StageExpansionButton {...stageExpansionButtonProps} />);
    expect(stageExpansionButtonProps.toggle).toHaveBeenCalledTimes(0);
    fireEvent.click(screen.getByRole('button'));
    expect(stageExpansionButtonProps.toggle).toHaveBeenCalledTimes(1);
  });
});
