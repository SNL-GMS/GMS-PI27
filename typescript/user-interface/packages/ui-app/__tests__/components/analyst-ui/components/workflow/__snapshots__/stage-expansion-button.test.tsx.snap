// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stage Expansion Button matches snapshot 1`] = `
<body>
  <div>
    <button
      class="bp5-button stage-row__expand-button"
      type="button"
    >
      <span
        aria-hidden="true"
        class="bp5-icon bp5-icon-standard bp5-icon-small-plus"
        data-icon="small-plus"
      />
    </button>
  </div>
</body>
`;
